File: ruby/controllers.rb


module ControllerOps
  def foo(param)
    @dev.op1(param * param)
  end
end


class ControllerA
  include ControllerOps

  def initialize(dev)
    @dev = dev
  end

  def run(param)
    @dev.op2(param)
  end
end


class ControllerB
  include ControllerOps

  def initialize(dev)
    @dev = dev
  end

  def run(param)
    @dev.op2(param, param)
  end
end

File: rust/src/controllers.rs

use soup;


pub struct Controller<'a, T> where T: 'a + soup::Device {
    dev: &'a mut T,
}

impl<'a, T> Controller<'a, T> where T: soup::Device {
    fn foo(&mut self, param: i32) {
        self.dev.op1(param * param)
    }
}


pub struct ControllerA<'a> {
    base: Controller<'a, soup::DeviceA>,
}


impl<'a> ControllerA<'a> {
    pub fn new(dev: &mut soup::DeviceA) -> ControllerA {
        ControllerA { dev: dev }
    }

    // pub fn foo(&mut self, param: i32) {
    //     self.dev.op1(param * param)
    // }

    pub fn run(&mut self, param: i32) -> i32 {
        self.dev.op2(param)
    }
}


pub struct ControllerB<'a> {
    dev: &'a mut soup::DeviceB,
}

impl<'a> ControllerB<'a> {
    pub fn new(dev: &mut soup::DeviceB) -> ControllerB {
        ControllerB { dev: dev }
    }

    pub fn foo(&mut self, param: i32) {
        self.dev.op1(param * param)
    }

    pub fn run(&mut self, param: i32) -> i32 {
        self.dev.op2(param, param)
    }
}
